crest/Assets/Crest/Crest/Scripts/Shapes/ShapeGerstner.cs

```diff
@@ -0,0 +1,619 @@
+﻿// Crest Ocean System
+
+// This file is subject to the MIT License as seen in the root of this folder structure (LICENSE)
+
+using UnityEngine;
+using UnityEngine.Experimental.Rendering;
+using UnityEngine.Rendering;
+using Unity.Collections.LowLevel.Unsafe;
+
+#if UNITY_EDITOR
+using UnityEditor;
+#endif
+
+namespace Crest
+{
+    /// <summary>
+    /// Gerstner ocean waves.
+    /// </summary>
+    [ExecuteAlways]
+    public partial class ShapeGerstner : MonoBehaviour, IFloatingOrigin
+    {
+        [Header("Wave Settings")]
+        [Tooltip("The spectrum that defines the ocean surface shape. Assign asset of type Crest/Ocean Waves Spectrum.")]
+        public OceanWaveSpectrum _spectrum;
+
+        [Tooltip("If a spectrum will not change at runtime, set this true to calculate the wave data once on first update rather than each frame."), SerializeField]
+        bool _spectrumIsStatic = true;
+
+        [Tooltip("Wind direction (angle from x axis in degrees)"), Range(-180, 180)]
+        public float _windDirectionAngle = 0f;
+        public Vector2 WindDir => new Vector2(Mathf.Cos(Mathf.PI * _windDirectionAngle / 180f), Mathf.Sin(Mathf.PI * _windDirectionAngle / 180f));
+
+        [Tooltip("Multiplier for these waves to scale up/down."), Range(0f, 1f)]
+        public float _weight = 1f;
+
+        [Tooltip("How much these waves respect the shallow water attenuation setting in the Animated Waves Settings. Set to 0 to ignore shallow water."), SerializeField, Range(0f, 1f)]
+        float _respectShallowWaterAttenuation = 1f;
+
+        [Header("Generation Settings")]
+        [Delayed, Tooltip("How many wave components to generate in each octave.")]
+        public int _componentsPerOctave = 8;
+
+        [Tooltip("Change to get a different set of waves.")]
+        public int _randomSeed = 0;
+
+        [Tooltip("Resolution to use for wave generation buffers. Low resolutions are more efficient but can result in noticeable patterns in the shape."), Delayed]
+        public int _resolution = 32;
+
+        [Tooltip("In Editor, shows the wave generation buffers on screen."), SerializeField]
+        bool _debugDrawSlicesInEditor = false;
+
+        Mesh _meshForDrawingWaves;
```

Comment:
Should this be serialised?

Reply:
Well spotted. I didnt explain this.

This is currently internal to the script and not serialised. I have downstream changes to generate the mesh from a spline, in which case the spline would set this value.

This plan does not expose this field to the user, so they won't be able to provide their own meshes, currently. I have the feeling that no one uses this functionality (the required gerstner shader to render waves weighted by vertex colour has never been requested. it was added recently by me after i recommended it to someone, and they did not use it in the end).

The spline stuff that is coming later is quite powerful (can create lots of interesting wave areas). Given this i propose not exposing this to the user on this script - for now at least. Let me know what you think.

Reply:
Makes perfect sense.